File: services/python-rag/app/rag/vector_store.py
import math

from app.rag.models import Chunk, SearchResult, VectorRecord


class InMemoryVectorStore:
    def __init__(self) -> None:
        self._records: dict[str, VectorRecord] = {}

    def upsert_many(self, chunks: list[Chunk], embeddings: list[list[float]]) -> None:
        for chunk, embedding in zip(chunks, embeddings, strict=True):
            self._records[chunk.id] = VectorRecord(chunk=chunk, embedding=embedding)

    def delete_by_document(self, document_id: str) -> None:
        chunk_ids = [
            chunk_id
            for chunk_id, record in self._records.items()
            if record.chunk.document_id == document_id
        ]
        for chunk_id in chunk_ids:
            del self._records[chunk_id]

    def search(
        self,
        query_embedding: list[float],
        top_k: int,
        filters: dict[str, object] | None = None,
    ) -> list[SearchResult]:
        filters = filters or {}
        scored: list[SearchResult] = []
        for record in self._records.values():
            if not self._matches_filters(record.chunk, filters):
                continue
            score = cosine_similarity(query_embedding, record.embedding)
            scored.append(SearchResult(chunk=record.chunk, score=score))

        scored.sort(key=lambda result: result.score, reverse=True)
        return scored[:top_k]

    def count_for_document(self, document_id: str) -> int:
        return sum(1 for record in self._records.values() if record.chunk.document_id == document_id)

    def _matches_filters(self, chunk: Chunk, filters: dict[str, object]) -> bool:
        for key, expected in filters.items():
            if chunk.metadata.get(key) != expected:
                return False
        return True


def cosine_similarity(left: list[float], right: list[float]) -> float:
    if len(left) != len(right):
        return 0.0

    dot = sum(a * b for a, b in zip(left, right, strict=True))
    left_length = math.sqrt(sum(a * a for a in left))
    right_length = math.sqrt(sum(b * b for b in right))
    if left_length == 0 or right_length == 0:
        return 0.0
    return dot / (left_length * right_length)

File: services/go-rag/internal/rag/vector_store.go
package rag

import (
	"math"
	"sort"
	"sync"
)

type vectorStore struct {
	mu     sync.RWMutex
	chunks map[string]Chunk
}

func newVectorStore() *vectorStore {
	return &vectorStore{
		chunks: make(map[string]Chunk),
	}
}

func (s *vectorStore) replaceDocumentChunks(documentID string, chunks []Chunk) {
	s.mu.Lock()
	defer s.mu.Unlock()

	for id, chunk := range s.chunks {
		if chunk.DocumentID == documentID {
			delete(s.chunks, id)
		}
	}
	for _, chunk := range chunks {
		chunk.Metadata = cloneStringMap(chunk.Metadata)
		chunk.Embedding = cloneFloatSlice(chunk.Embedding)
		s.chunks[chunk.ID] = chunk
	}
}

func (s *vectorStore) search(queryEmbedding []float64, topK int, filters map[string]string) []RetrievedChunk {
	s.mu.RLock()
	defer s.mu.RUnlock()

	if topK <= 0 {
		topK = 5
	}

	results := make([]RetrievedChunk, 0, len(s.chunks))
	for _, chunk := range s.chunks {
		if !metadataMatches(chunk.Metadata, filters) {
			continue
		}

		results = append(results, RetrievedChunk{
			ChunkID:       chunk.ID,
			DocumentID:    chunk.DocumentID,
			DocumentTitle: chunk.Metadata["document_title"],
			Ordinal:       chunk.Ordinal,
			Content:       chunk.Content,
			Metadata:      cloneStringMap(chunk.Metadata),
			Score:         cosine(queryEmbedding, chunk.Embedding),
		})
	}

	sort.Slice(results, func(i int, j int) bool {
		if results[i].Score == results[j].Score {
			return results[i].ChunkID < results[j].ChunkID
		}
		return results[i].Score > results[j].Score
	})

	if len(results) > topK {
		results = results[:topK]
	}
	return results
}

func cosine(left []float64, right []float64) float64 {
	limit := len(left)
	if len(right) < limit {
		limit = len(right)
	}

	var dot float64
	var leftMagnitude float64
	var rightMagnitude float64
	for i := 0; i < limit; i++ {
		dot += left[i] * right[i]
		leftMagnitude += left[i] * left[i]
		rightMagnitude += right[i] * right[i]
	}
	if leftMagnitude == 0 || rightMagnitude == 0 {
		return 0
	}
	return dot / (math.Sqrt(leftMagnitude) * math.Sqrt(rightMagnitude))
}

func metadataMatches(metadata map[string]string, filters map[string]string) bool {
	for key, expected := range filters {
		if metadata[key] != expected {
			return false
		}
	}
	return true
}
